File: server/pkg/request/request.go
package request

import (
	"errors"
	"time"

	"golang.org/x/exp/slices"
)

var (
	ErrEmptyItems = errors.New("items cannot be empty")
	ErrEmptyTitle = errors.New("title cannot be empty")
)

type Request struct {
	id          ID
	workspace   WorkspaceID
	project     ProjectID
	items       ItemList
	title       string
	description string
	createdBy   UserID
	reviewers   UserIDList
	state       State
	updatedAt   time.Time
	approvedAt  *time.Time
	closedAt    *time.Time
	thread      ThreadID
}

func (r *Request) ID() ID {
	return r.id
}

func (r *Request) Workspace() WorkspaceID {
	return r.workspace
}

func (r *Request) Project() ProjectID {
	return r.project
}

func (r *Request) Items() ItemList {
	return slices.Clone(r.items)
}

func (r *Request) Title() string {
	return r.title
}

func (r *Request) Description() string {
	return r.description
}

func (r *Request) CreatedBy() UserID {
	return r.createdBy
}

func (r *Request) Reviewers() UserIDList {
	return r.reviewers
}

func (r *Request) State() State {
	return r.state
}

func (r *Request) CreatedAt() time.Time {
	return r.id.Timestamp()
}

func (r *Request) UpdatedAt() time.Time {
	return r.updatedAt
}

func (r *Request) ApprovedAt() *time.Time {
	return r.approvedAt
}

func (r *Request) ClosedAt() *time.Time {
	return r.closedAt
}

func (r *Request) Thread() ThreadID {
	return r.thread
}

func (r *Request) SetTitle(title string) error {
	if title == "" {
		return ErrEmptyTitle
	}
	r.title = title
	return nil
}

func (r *Request) SetDescription(description string) {
	r.description = description
}

func (r *Request) SetReviewers(reviewers []UserID) {
	r.reviewers = reviewers
}

func (r *Request) SetItems(items ItemList) {
	r.items = slices.Clone(items)
}

func (r *Request) SetState(state State) {
	r.state = state
}

File: web/src/gql/fragments/request.ts
import { gql } from "@apollo/client";

import { threadFragment } from "@reearth-cms/gql/fragments";

export const requestFragment = gql`
  fragment requestFragment on Request {
    id
    items {
      itemId
      version
      ref
      item {
        version
        parents
        refs
        value {
          id
          title
          schemaId
          modelId
          model {
            name
          }
          fields {
            schemaFieldId
            type
            value
            itemGroupId
          }
          referencedItems {
            id
            title
            schemaId
            createdBy {
              ... on Integration {
                name
              }
              ... on User {
                name
              }
            }
            status
            version
            createdAt
            updatedAt
          }
          schema {
            id
            fields {
              id
              type
              title
              key
              description
              required
              unique
              isTitle
              multiple
              typeProperty {
                ... on SchemaFieldText {
                  defaultValue
                  maxLength
                }
                ... on SchemaFieldTextArea {
                  defaultValue
                  maxLength
                }
                ... on SchemaFieldMarkdown {
                  defaultValue
                  maxLength
                }
                ... on SchemaFieldAsset {
                  assetDefaultValue: defaultValue
                }
                ... on SchemaFieldSelect {
                  selectDefaultValue: defaultValue
                  values
                }
                ... on SchemaFieldInteger {
                  integerDefaultValue: defaultValue
                  min
                  max
                }
                ... on SchemaFieldBool {
                  defaultValue
                }
                ... on SchemaFieldURL {
                  defaultValue
                }
                ... on SchemaFieldReference {
                  modelId
                }
                ... on SchemaFieldGroup {
                  groupId
                }
              }
            }
          }
        }
      }
    }
    title
    description
    createdBy {
      id
      name
      email
    }
    workspaceId
    projectId
    threadId
    reviewersId
    state
    createdAt
    updatedAt
    approvedAt
    closedAt
    thread {
      ...threadFragment
    }
    project {
      id
      name
      createdAt
      updatedAt
    }
    reviewers {
      id
      name
      email
    }
  }
  ${threadFragment}
`;

export default requestFragment;
